### PR TITLE
Fix signed-unsigned warnings

### PR DESCRIPTION
**Pull request by malfet**
And few typos
Test Plan: CI

### CODE REVIEW

**Comment by ezyang** [comment]
@zdevito can you take a look at this? I fear that manually fixing these warnings is a losing battle unless we come up with some broader strategy about what to do in these cases. In the past, we've forced people to use `int64_t` as much as possible, which has helped, but whenever we call standard library functions like `size()` you end up with unsigned creeping in.

(The patch looks generally reasonable, and we don't have to block the PR itself on fixing htese things.)

**Comment by malfet** [comment]
@ezyang would it be easier if I split this PR into few smaller ones?

File: caffe2/operators/boolean_mask_ops.cc
Comment (zdevito) [inline caffe2/operators/boolean_mask_ops.cc]: This also changes the size to 64-bit. We should use uint32_t here, otherwise there might be performance differences.

File: caffe2/operators/byte_weight_dequant_op.h
Comment (zdevito) [inline caffe2/operators/byte_weight_dequant_op.h]: same here, it was 32-bit before.

File: caffe2/operators/elementwise_ops.cc
Comment (zdevito) [inline caffe2/operators/elementwise_ops.cc]: Same here

File: caffe2/operators/index_ops.h
Comment (zdevito) [inline caffe2/operators/index_ops.h]: same here, not going to mark the rest

**Review by zdevito — COMMENTED** [review]
I think everything looks fine here except that we should be replacing int with uint32_t and not size_t. The point is to stop the warnings, I don't want to also change indexing math to 64-bits in a way that might slow something down.

**Comment by zdevito** [comment]
re warnings (not specific to this PR): the best bet is to actually get warnings clean by turning off the warnings our code is not clean against, and then gradually expand the set. Currently we just introduce warnings because they are impossible to see.

**Comment by ezyang** [comment]
@zdevito IIUC, the 32-bit/64-bit difference isn't a big deal in CPU code; it's just in CUDA where it really slays you?

File: caffe2/operators/boolean_mask_ops.cc
Comment (malfet) [reply caffe2/operators/boolean_mask_ops.cc]: Actually, in  this case, it looks more reasonable to change `M` and `N` types from `size_t` to `int`

**Comment by malfet** [comment]
Updated the diff to use `auto foo = 0U` instead of `int foo = 0` or `size_t foo = 0U`. This way compiler is  free to choose whatever type it thinks fits best for the architecture, which is currently `unsigned int` for all architectures we support.

**Comment by malfet** [comment]
@ezyang, I don't think x86_64 is capable of non-vectorized 32-bit arithmetic.
IMO performance would only be affected on 32-bit architectures (ARMv7 or x86), GPUs or if value ever needs to be stored to memory.

**Review by ezyang — APPROVED** [review]

File: caffe2/contrib/gloo/allreduce_ops.h
Comment (BIT-silence) [inline caffe2/contrib/gloo/allreduce_ops.h]: Personally I prefer for (size_t i = 0; ...), but this is fine.

Comment (jerryzh168) [reply caffe2/contrib/gloo/allreduce_ops.h]: yeah I think size_t is better, I'll change it after my stack of diffs land

Comment (jerryzh168) [reply caffe2/contrib/gloo/allreduce_ops.h]: I mean I'll change this in my code

Comment (malfet) [reply caffe2/contrib/gloo/allreduce_ops.h]: Please loop @zdevito on that one, because I had `size_t` there initially and he suggested to use `unsigned int` out of performance considerations.

Comment (malfet) [reply caffe2/contrib/gloo/allreduce_ops.h]: But personally I agree, that if upper limit it 64-bit unsigned value, then loop iterator should also be 64-bit unsigned value.

Comment (BIT-silence) [reply caffe2/contrib/gloo/allreduce_ops.h]: Actually all of params.inputs, devices_ or elements are std::vectors. While in std::vector the return type of std::vector<T>::size() is std::vector<T>::size_type, which in most cases is size_t. If there is platform concerns, I think we'd better use std::vector<T>::size_type i = 0; And I don't think there should be performance issues for such a for-loop iterative variable especially the vectors should be considerable small. 
And for the case in the computation code, do we have any benchmark for the differences between the type in for-loops?

**Pull request closed without merging**
